File: ruby/20_valid_parentheses.rb
# @param {String} s
# @return {Boolean}


# 给定字符串由三种括号组成，判断是否合法，判断标准为括号两两闭合顺序正确同时必须同一种括号
# 可以使用栈，当括号为前一部分时，入栈；为后一部分时取top进行比较看是否为同一类型，否则返回false
# 最后看栈是否为空即可

def is_valid(s)
	stack = []

	s.each_char do |c|
		case c
		when '(', '[', '{'
			stack.push(c)
		when ')'
			return false if stack.pop != '('
		when ']'
			return false if stack.pop != '['
		when '}'
			return false if stack.pop != '{'
		end
	end

	stack.empty?
end

File: cpp/20_valid_parentheses.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

/*
 给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。

有效字符串需满足：

左括号必须用相同类型的右括号闭合。
左括号必须以正确的顺序闭合。
*/

class Solution {
public:
    bool isValid(string s) {
		int nsize = s.size();
		if (nsize < 2) { return false; }

		stack<char> stk;
		for (auto ch : s) {
			if (ch == '(' || ch == '[' || ch == '{') {
				stk.push(ch);
			} else {
				// case ')'
                if (stk.empty()) { return false; }

				char top = stk.top(); stk.pop();
				if (ch == ')') {
					if (top != '(') { return false; }
				}
				if (ch == ']') {
					if (top != '[') { return false; }
				}
				if (ch == '}') {
					if (top != '{') { return false; }
				}
			}
		}

		// case '(('
		return stk.empty();
    }
};
